pkgs/development/compilers/julia/0.7.nix

```diff
@@ -0,0 +1,200 @@
+  suitesparse = suitesparse_.override { inherit openblas; };
+in
+
+let
+  dsfmtVersion = "2.2.3";
+  dsfmt = fetchurl {
+    url = "http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/SFMT/dSFMT-src-${dsfmtVersion}.tar.gz";
+    sha256 = "03kaqbjbi6viz0n33dk5jlf6ayxqlsq4804n7kwkndiga9s4hd42";
+  };
+
+  libuvVersion = "be317349252699670131395f125c3861d793ca86";
+  libuv = fetchurl {
+    url = "https://api.github.com/repos/JuliaLang/libuv/tarball/${libuvVersion}";
+    sha256 = "0s6dik373yvpj8xxmps3wasr6s5wrzbccpfw7vzlclbgxlpyjq0i";
+  };
+
+  rmathVersion = "0.1";
+  rmath-julia = fetchurl {
+    url = "https://api.github.com/repos/JuliaLang/Rmath-julia/tarball/v${rmathVersion}";
+    sha256 = "1qyps217175qhid46l8f5i1v8i82slgp23ia63x2hzxwfmx8617p";
+  };
+
+  libwhichVersion = "81e9723c0273d78493dc8c8ed570f68d9ce7e89e";
+  libwhich = fetchurl {
+    url = "https://api.github.com/repos/vtjnash/libwhich/tarball/${libwhichVersion}";
+    sha256 = "1p7zg31kpmpbmh1znrk1xrbd074agx13b9q4dcw8n2zrwwdlbz3b";
+  };
+
+
+  virtualenvVersion = "15.0.0";
+  virtualenv = fetchurl {
+    url = "mirror://pypi/v/virtualenv/virtualenv-${virtualenvVersion}.tar.gz";
+    sha256 = "06fw4liazpx5vf3am45q2pdiwrv0id7ckv7n6zmpml29x6vkzmkh";
+  };
+in
+
+stdenv.mkDerivation rec {
+  pname = "julia";
+  version = "0.7.0";
+  name = "${pname}-${version}";
+
+  src = fetchFromGitHub {
+    owner = "JuliaLang";
+    repo = "${pname}";
+    rev = "v${version}-alpha";
+    sha256 = "1w29knnkm841s84bkx0ksxr81j0xwpvz1vdxf4xcdk18v3wc2spc";
+  };
+
+  prePatch = ''
+    mkdir deps/srccache
+    cp "${libuv}" "./deps/srccache/libuv-${libuvVersion}.tar.gz"
+    cp "${dsfmt}" "./deps/srccache/dsfmt-${dsfmtVersion}.tar.gz"
+    cp "${rmath-julia}" "./deps/srccache/Rmath-julia-${rmathVersion}.tar.gz"
+    cp "${virtualenv}" "./deps/srccache/virtualenv-${virtualenvVersion}.tar.gz"
+    cp "${libwhich}" "./deps/srccache/libwhich-${libwhichVersion}.tar.gz"
+  '';
+
+  patches = [
+    ./0001.1-use-system-utf8proc.patch
+    ./0002-disable-network-in-tests.patch
+    ./0003-disable-Distributed.patch
+    # ./0002-use-system-suitesparse.patch
+  ] ++ stdenv.lib.optional stdenv.needsPax ./0004-hardened.patch;
+
+  postPatch = ''
+    patchShebangs . contrib
+  '';
+
+  buildInputs = [
+    arpack fftw fftwSinglePrec gmp libgit2 libunwind mpfr
+    pcre2.dev openblas openlibm openspecfun readline suitesparse utf8proc
+    zlib llvm
+  ]
+  ++ stdenv.lib.optionals stdenv.isDarwin [CoreServices ApplicationServices]
+  ;
+
+  nativeBuildInputs = [ curl gfortran m4 makeWrapper patchelf perl python2 which ]
+    ++ stdenv.lib.optional stdenv.needsPax paxctl;
+
+  makeFlags =
+    let
+      arch = head (splitString "-" stdenv.system);
+      march = { "x86_64" = "x86-64"; "i686" = "pentium4"; }."${arch}"
+              or (throw "unsupported architecture: ${arch}");
+      # Julia requires Pentium 4 (SSE2) or better
+      cpuTarget = { "x86_64" = "x86-64"; "i686" = "pentium4"; }."${arch}"
```

Comment:
You don't actually need to set CPU target if it's identical to `MARCH`. However, it would be a good idea to set it to something more fancy so that several images are built, to enable the use of recent instructions where possible (this matters a lot for scientific work). Upstream [uses this](https://github.com/JuliaCI/julia-buildbot/blob/49bde486c0412fad0969f46fb58b621f33caa744/master/inventory.py#L61-L73):
- x86: `'JULIA_CPU_TARGET="pentium4;sandybridge,-xsaveopt,clone_all" '`
- x86_64: `'JULIA_CPU_TARGET="generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)'`

Reply:
@7c6f434c I've taken it from your commit, do you know why/if it's still required? I don't like messing with this because years back I had non-deterministic builds (because of SSE2 optimizations https://github.com/NixOS/nixpkgs/issues/13662)

Reply:
Yeah, but Julia 0.7 is much more robust now, it will always use SSE2 and it will throw an error if compiler options are incorrect. Since the `JULIA_CPU_TARGET` values I posted come from upstream, they should be safe to use.

Reply:
I think it was way before 0.7, though, if there are improvements. we could try being optimistic.

Reply:
So like this?

```nix
      cpuTarget = { "x86_64" = "generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)";
                    "i686" = "pentium4;sandybridge,-xsaveopt,clone_all"; }."${arch}"
```

(Compiling)